funsor/tensor.py

```diff
@@ -381,6 +381,20 @@ def materialize(self, x):
         subs = tuple(subs)
         return substitute(x, subs)
 
+    @property
+    def backend(self):
+        if torch.is_tensor(self.data):
+            return "torch"
+        else:
+            return "numpy"
+
+    @property
+    def logsumexp_backend(self):
```

Comment:
nit: I think this is better as a global dict accessible to `_eager_contract_tensors()`:
```py
# in funsor/cnf.py
BACKEND_TO_LOGSUMEXP_BACKEND = {
    "torch": "pyro.ops.einsum.torch_log",
    "numpy": "funsor.einsum.numpy_log",
}

@eager.register(Contraction, ops.LogAddExpOp, ops.AddOp, frozenset, Tensor, Tensor)
def eager_contraction_tensor(red_op, bin_op, reduced_vars, *terms):
    if not all(term.dtype == "real" for term in terms):
        raise NotImplementedError('TODO')
    backend = BACKEND_TO_LOGSUMEXP_BACKEND[terms[0].backend]
    return _eager_contract_tensors(reduced_vars, terms, backend=backend)
```

Reply:
Thanks!